### DISCUSSION TITLE
How to configure electrs to retrieve large address histories

### DISCUSSION THREAD

**Discussion by cricktor** [romanz/electrs#472] [General]
Hi,
I'm running Bitcoin Core v0.21.1 and electrs with Raspiblitz v1.7.0 as my personal node and Electrum server on a Raspi 4B with 8GB RAM.

Sometimes I like to do some blockchain digging and analysis and run into problems when a big address history needs to be reported back to the Electrum client or local Bitcoinexplorer/mempool.space.

I tried to find some configuration clues, but wasn't very lucky. The local installed tools as like bitcoinexplorer.org and mempool.space request the address history from the electrs server. (It's configured to do it.)

I don't mind if the Raspi has to do some deep data digging, it should've enough RAM. Can I configure electrs such that it would retrieve big address histories like some thousands of transactions?

Any advise is very appreciated! If I missed to provide some helpful further details, let me know I will amend it.
bitcoind has txindex and blockfilterindex active and is of course fully synched.

**Comment by cricktor** [comment]
In ~electrs/.electrs/config.toml I bumped txid_limit to 100000 or more but I still have timeout issues with official Electrum v4.1.5 wallet when it tries to synchronize a watch only wallet of some bitcoin burn addresses that I have collected (don't have some better address examples right now; those burn addresses don't have such a huge transaction history, so I don't really understand why it causes such sync issues).
My local Electrum wallet synchronized only with my own electrs server.

**Comment by Kixunil** [comment]
Hi, the current Electrum protocol is not great when it comes to queries with many results, so we will have to wait for the new proposal to add pagination. In the meantime consider trying `p2p` branch which contains some optimizations and may work better. It'll become a new version somewhat soon anyway.

> I tried to find some configuration clues, but wasn't very lucky.

Did you read the documentation? It explicitly points out there's `--help` page.

**Reply by romanz** [reply]
You can increase the [default limit](https://github.com/romanz/electrs/blob/b916c2802bdee5b9c98b420d2a93dada3cdf055e/internal/config_specification.toml#L96) using the `--index-lookup-limit` flag (currently supported on `p2p` branch):
```
electrs --network bitcoin --db-dir ./db --daemon-dir ~/.bitcoin --index-lookup-limit 1000
```

**Comment by romanz** [comment]
[Raspiblitz 1.7.1](https://github.com/rootzoll/raspiblitz/blob/v1.7/CHANGES.md#whats-new-in-version-171-of-raspiblitz) runs electrs 0.9.0, which should have better performance :)

**Comment by MCM-Mike** [comment]
Is there an update on:

```
You can increase the [default limit](https://github.com/romanz/electrs/blob/b916c2802bdee5b9c98b420d2a93dada3cdf055e/internal/config_specification.toml#L96) using the --index-lookup-limit flag (currently supported on p2p branch):

electrs --network bitcoin --db-dir ./db --daemon-dir ~/.bitcoin --index-lookup-limit 1000``` 

and pagination ?

**Reply by Kixunil** [reply]
What update do you expect on that option?

Sadly, pagination is stuck on protocol spec, no clue how to unstuck it. I'm tempted to say "let's implement it anyway" but it wouldn't make sense if some client doesn't implement it too. Maybe Sparrow could get onboard which seems more active in development than Electrum and more of their users seem to need this? What do you think @craigraw?

**Reply by MCM-Mike** [reply]
I was looking for a full history of transactions as well as balances, to my understanding the `index-lookup-limit` was a workaround to archive this when you had a full indexed bitcoin-node.

**Reply by Kixunil** [reply]
So you have a large number of transactions on a single address and you want to fetch all of them?

**Reply by MCM-Mike** [reply]
>So you have a large number of transactions on a single address and you want to fetch all of them?
Yes.

How is `electrs` acting when looking up multiple addresses concurrent or at the same time with a lot of transaction history ? 
I am asking because I had some issues before moving to `electrs`.

**Reply by cricktor** [reply]
My use case is that sometimes I do some blockchain research for fun or when I get my nose pointed to some interesting transactions or addresses. Sometimes this turned out to be frustrating with the base setup of my RaspiBlitz node because if an address or collection of addresses has a large transaction history, it failed to deliver a complete result or any at all. (And I prefer to use the digging with my local node, not only for my personal wallets, transactions and addresses.)

Usually I point my Electrum wallets to the electrs of RaspiBlitz. Most of the time I use the block explorers within the local versions of mempool.space and RPC-BTC-Explorer (equivalent to bitcoinexplorer.org). Both blockchain explorers complain pretty fast, if you hit a transaction history of a few thousands. And to my knowledge this isn't an issue of the block explorers. The observed error blames electrs (or better the lack of a suitable configuration of it, I guess).

I don't mind to wait if it takes some time for a large address history. All I want is to get the full "story" even when my RPi (8GB RAM) has quite some work to fulfill.

I admit that I still have to dig through options and documentation to find a suitable setup. Any hints are welcome!

**Reply by Kixunil** [reply]
@MCM-Mike sadly large requests are not great and there are all kinds of issues. Reusing a single address is a bad practice in Bitcoin anyway, so I don't recommend that and electrs is not optimized for it. Multiple addresses should be doable with multiple requests but it may take some time.

@cricktor RPi will probably deliver poor performance regardless of implementation but unless it's single address-many transactions I wouldn't attribute fault entirely to electrs even though it could've been faster.

> The observed error blames electrs (or better the lack of a suitable configuration of it, I guess).

Could you share such error message?

> I don't mind to wait if it takes some time for a large address history.

This actually points to the fault of explorers being inpatient. They should at least have an option to bump the timeout.

**Reply by cricktor** [reply]
@Kixunil Take e.g. Satoshi's receiving address of the Genesis block: it received a lot of donation and dust transactions.
Response from mempool.space explorer is:

> Address
[1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa](http://raspiblitz.local:4080/address/1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa)

>  
> Error loading address data.
> The number of transactions on this address exceeds the Electrum server limit

>  
> Consider viewing this address on the official Mempool website instead:
> https://mempool.space/address/1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa
> http://mempoolhqx4isw62xs7abwphsq7ldayuidyx2v2oethdhhj6mlo2r6ad.onion/address/1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa

>  
> (>200 index entries, query may take too long)


RPC-BTC-Explorer mumbles:
```
Error retrieving transaction history (#1)
{
    "e": {
        "code": 1,
        "message": ">200 index entries, query may take too long",
        "userData": {
            "address": "1A1zP1eP5QGefi2DMPTfTL5SLmv7DivfNa",
            "sort": "desc",
            "limit": 10,
            "offset": 0
        }
    },
    "status": "rejected"
}
```

**Reply by MCM-Mike** [reply]
>... Multiple addresses should be doable with multiple requests but it may take some time.

Thanks, I do know about best practice behavior for BTC addresses.
But leading with a "good" example are exchanges who are reusing same "deposit" addresses for their clients, etc  and sometimes even "withdraw" addresses. This is why it's sometimes important to have all transaction's of a single address (native)-segwit, legacy and now taproot).

_(could lead a bit OT)_
@cricktor can you use "electrs" with btc-rpc-explorer, what are your settings there in terms of `enviroment` variables for `BTCEXP_ADDRESS_API` ?

**Reply by Kixunil** [reply]
I see, the error message is reasonable and yes, it's the lack of a suitable configuration of electrs.

Frankly, I don't think it's much of a privacy issue to lookup these public addresses in public explorers. Even if an explorer was selling the data of customers to chain analysis (like already happened) if they are stupid enough to think it's your address it actually helps the ecosystem.

Anyway, exchanges reusing addresses is nuisance that's true. However maybe not optimizing it puts pressure on them to stop doing stupid shit?

**Reply by cricktor** [reply]
> ... can you use "electrs" with btc-rpc-explorer, what are your settings there in terms of `enviroment` variables for `BTCEXP_ADDRESS_API` ?

Likely relevant settings in RaspiBlitz's node configuration `env`-file for btc-rpc-explorer:
```
BTCEXP_BITCOIND_RPC_TIMEOUT=10000
BTCEXP_ADDRESS_API=electrumx
BTCEXP_ELECTRUMX_SERVERS=tcp://127.0.0.1:50001
```

> Frankly, I don't think it's much of a privacy issue to lookup these public addresses in public explorers. Even if an explorer was selling the data of customers to chain analysis (like already happened) if they are stupid enough to think it's your address it actually helps the ecosystem.

If I'm dealing with addresses that are not related to me in any way, I wouldn't mind using suitable public explorers. But what kind if argument is yours? If I want to keep privacy for blockchain analysis which does touch my own addresses, I need my local node and tools setup to be able to deliver the data properly. And this includes a suitable configuration for electrs that delivers in the best case any address history, even if it takes some time and delays. If electrs is not the tool for this, I'm fine with that: then I need to find a more suitable tool. If I need bigger indices for electrs to solve it, I'm fine with that too, then it's my decission to make it work.

Currently I'm just not yet aware of where and what to tweak.

**Reply by Kixunil** [reply]
Yes, use your own node for your own stuff. If you don't reuse addresses (which would undermine your goal anyway) electrs should be fine.

You may want to look into blockstream for of electrs if you want to reuse addresses. It's more performant but eats much more space.
